streams/src/test/java/org/apache/kafka/streams/smoketest/SmokeTestDriver.java

```diff
@@ -0,0 +1,492 @@
+        SmokeTestClient streams4 = new SmokeTestClient(createDir(stateDir, "4"), kafka, zookeeper);
+
+        System.out.println("starting the driver");
+        driver.start();
+
+        System.out.println("starting the first and second client");
+        streams1.start();
+        streams2.start();
+
+        sleep(10000);
+
+        System.out.println("starting the third client");
+        streams3.start();
+
+        System.out.println("closing the first client");
+        streams1.close();
+        System.out.println("closed the first client");
+
+        sleep(10000);
+
+        System.out.println("starting the forth client");
+        streams4.start();
+
+        driver.join();
+
+        System.out.println("driver stopped");
+
+        streams2.close();
+        streams3.close();
+        streams4.close();
+
+        System.out.println("shutdown");
+    }
+
+    public static Map<String, Set<Integer>> generate(String kafka, final int numKeys, final int maxRecordsPerKey) throws Exception {
+        Properties props = new Properties();
+        props.put(ProducerConfig.CLIENT_ID_CONFIG, "SmokeTest");
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class);
+
+        KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(props);
+
+        int numRecordsProduced = 0;
+
+        Map<String, Set<Integer>> allData = new HashMap<>();
+        ValueList[] data = new ValueList[numKeys];
+        for (int i = 0; i < numKeys; i++) {
+            data[i] = new ValueList(i, i + maxRecordsPerKey - 1);
+            allData.put(data[i].key, new HashSet<Integer>());
+        }
+        Random rand = new Random();
+
+        int remaining = data.length;
+
+        while (remaining > 0) {
+            int index = rand.nextInt(remaining);
+            String key = data[index].key;
+            int value = data[index].next();
+
+            if (value < 0) {
+                remaining--;
+                data[index] = data[remaining];
+                value = END;
+            }
+
+            if (value != -1) {
+                ProducerRecord<byte[], byte[]> record =
+                        new ProducerRecord<>("data", stringSerializer.serialize("", key), integerSerializer.serialize("", value));
+
+                producer.send(record);
+
+                if (value != END) {
+                    numRecordsProduced++;
+                    allData.get(key).add(value);
+
+                    if (numRecordsProduced % 100 == 0)
+                        System.out.println(numRecordsProduced + " records produced");
+
+                    Thread.sleep(10);
+                }
+            }
+        }
+
+        producer.close();
+
+        return Collections.unmodifiableMap(allData);
+    }
+
+    private static void shuffle(int[] data, int windowSize) {
```

Comment:
Brief comment would be useful to indicate why you provided a window for "semi-correlated shuffling"